allennlp/modules/maxout.py

```diff
@@ -0,0 +1,98 @@
+        ``num_layers``.
+    pool_sizes : ``Union[int, Sequence[int]]``
+        The size of max-pools.  If this is a single ``int``, we use
+        it for all maxout layers.  If it is a ``Sequence[int]``, ``len(pool_sizes)`` must be
+        ``num_layers``.
+    dropout : ``Union[float, Sequence[float]]``, optional
+        If given, we will apply this amount of dropout after each layer.  Semantics of ``float``
+        versus ``Sequence[float]`` is the same as with other parameters.
+    """
+    def __init__(self,
+                 input_dim: int,
+                 num_layers: int,
+                 output_dims: Union[int, Sequence[int]],
+                 pool_sizes: Union[int, Sequence[int]],
+                 dropout: Union[float, Sequence[float]] = 0.0) -> None:
+        super(Maxout, self).__init__()
+        if not isinstance(output_dims, list):
+            output_dims = [output_dims] * num_layers  # type: ignore
+        if not isinstance(pool_sizes, list):
+            pool_sizes = [pool_sizes] * num_layers  # type: ignore
+        if not isinstance(dropout, list):
+            dropout = [dropout] * num_layers  # type: ignore
+        if len(output_dims) != num_layers:
+            raise ConfigurationError("len(output_dims) (%d) != num_layers (%d)" %
+                                     (len(output_dims), num_layers))
+        if len(pool_sizes) != num_layers:
+            raise ConfigurationError("len(pool_sizes) (%d) != num_layers (%d)" %
+                                     (len(pool_sizes), num_layers))
+        if len(dropout) != num_layers:
+            raise ConfigurationError("len(dropout) (%d) != num_layers (%d)" %
+                                     (len(dropout), num_layers))
+
+        self._pool_sizes = pool_sizes
+        input_dims = [input_dim] + output_dims[:-1]
+        linear_layers = []
+        for layer_input_dim, layer_output_dim, pool_size in zip(input_dims, output_dims, pool_sizes):
+            linear_layers.append(torch.nn.Linear(layer_input_dim, layer_output_dim * pool_size))
+        self._linear_layers = torch.nn.ModuleList(linear_layers)
+        dropout_layers = [torch.nn.Dropout(p=value) for value in dropout]
+        self._dropout = torch.nn.ModuleList(dropout_layers)
+        self._output_dim = output_dims[-1]
+        self._input_dim = input_dim
+
+    def get_output_dim(self):
+        return self._output_dim
+
+    def get_input_dim(self):
+        return self._input_dim
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        # pylint: disable=arguments-differ
+        output = inputs
+        for layer, dropout, pool_size in zip(self._linear_layers, self._dropout, self._pool_sizes):
+            affine_output = layer(output)
+            # -1 in shape should be the output_dim of the maxout layer. This lets
+            # the maxout layer generalize the non-2D input.
+            maxed_output = torch.max(affine_output.view(*inputs.size()[:-1], -1, pool_size), dim=-1)[0]
```

Comment:
Preference for putting the shape computation on the line above this, and removing the -1, so it's more obvious what's going on here, and less likely to have bugs.  You have a comment that the "-1 in the shape should be the `output_dim`" - why not just use the `output_dim`, then, and remove the need for the comment, and make this crash if somehow there's an error there?  I had to think for a while about whether this was the right computation, and originally thought it was wrong, because the -1 makes it hard to reason about these things.

Reply:
Sorry for the acrobatics, I wanted to avoid storing the `output_dim` of each of the layers, but it seems worth it for simplicity.